src/mpid/ch4/netmod/ofi/util.c

```diff
@@ -60,51 +65,116 @@ int MPIDI_OFI_mr_key_allocator_init(void)
-            MPIDI_OFI_INDEX_CALC(val, nval, 4, 0xFULL);
-            MPIDI_OFI_INDEX_CALC(val, nval, 2, 0x3ULL);
-            nval -= val & 0x1ULL;
-            mr_key_allocator.bitmask[i] &= ~(0x1ULL << (nval - 1));
-            mr_key_allocator.last_free_mr_key = i;
-            return i * sizeof(uint64_t) * 8 + (nval - 1);
-        }
-        if (i == mr_key_allocator.num_ints - 1) {
-            mr_key_allocator.num_ints += mr_key_allocator.chunk_size;
-            mr_key_allocator.bitmask = MPL_realloc(mr_key_allocator.bitmask,
-                                                   sizeof(uint64_t) * mr_key_allocator.num_ints,
-                                                   MPL_MEM_RMA);
-            MPIR_Assert(mr_key_allocator.bitmask);
-            memset(&mr_key_allocator.bitmask[i + 1], 0xFF,
-                   sizeof(uint64_t) * mr_key_allocator.chunk_size);
-        }
+    uint64_t ret_key = MPIDI_OFI_INVALID_MR_KEY;
+    int id = key_type;
+
+    switch (key_type) {
+        case MPIDI_OFI_LOCAL_MR_KEY:
+            {
+                uint64_t i;
+                for (i = mr_key_allocator[id].last_free_mr_key; i < mr_key_allocator[id].num_ints;
+                     i++) {
+                    if (mr_key_allocator[id].bitmask[i]) {
+                        register uint64_t val, nval;
+                        val = mr_key_allocator[id].bitmask[i];
+                        nval = 2;
+                        MPIDI_OFI_INDEX_CALC(val, nval, 32, 0xFFFFFFFFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 16, 0xFFFFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 8, 0xFFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 4, 0xFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 2, 0x3ULL);
+                        nval -= val & 0x1ULL;
+                        mr_key_allocator[id].bitmask[i] &= ~(0x1ULL << (nval - 1));
+                        mr_key_allocator[id].last_free_mr_key = i;
+                        ret_key = i * sizeof(uint64_t) * 8 + (nval - 1);
+                    }
+                    if (i == mr_key_allocator[id].num_ints - 1) {
+                        mr_key_allocator[id].num_ints += mr_key_allocator[id].chunk_size;
+                        mr_key_allocator[id].bitmask = MPL_realloc(mr_key_allocator[id].bitmask,
+                                                                   sizeof(uint64_t) *
+                                                                   mr_key_allocator[id].num_ints,
+                                                                   MPL_MEM_RMA);
+                        MPIR_Assert(mr_key_allocator[id].bitmask);
+                        memset(&mr_key_allocator[id].bitmask[i + 1], 0xFF,
+                               sizeof(uint64_t) * mr_key_allocator[id].chunk_size);
+                    }
+                }
+                break;
+            }
+
+        case MPIDI_OFI_COLL_MR_KEY:
```

Comment:
So I'm thinking that the collective allocator can be even simpler than this. My idea is that it takes the context_id and places it in the range reserved for collective mr bits. Then the rest of the key is just a counter incremented for each registration. Basically:
```
win->comm->context_id + win->counter++
```
with protections to ensure the counter does not spill over into the collective bit range. Thoughts?

Reply:
a little more detailed example:
```
mr_key = win->comm->context_id << MPIDI_OFI_COLL_MR_SHIFT;
mr_key += win->counter++;
```
Do we need to actually return keys to allocator in this case? Or can we just allocate them consecutively on the window until the unlikely event that the value rolls over?